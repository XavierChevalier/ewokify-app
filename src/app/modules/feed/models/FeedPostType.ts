export enum AvailableFeedPostType {
  newSingle = 'newSingle',
  newAlbum = 'newAlbum',
}
