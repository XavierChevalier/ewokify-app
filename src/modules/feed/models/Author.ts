import { z } from 'zod'

export const Author = z.strictObject({
  name: z.string(),
  avatar: z.string().url(),
})

export type Author = z.infer<typeof Author>
