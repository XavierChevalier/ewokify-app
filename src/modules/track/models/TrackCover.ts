import { z } from 'zod'
import { TrackBands, TrackTitle } from '@/modules/track/models/Track'

export const TrackCover = z.string().url()
export type TrackCover = z.infer<typeof TrackCover>

export const TrackOfTrackCover = z.strictObject({
  cover: TrackCover,
  title: TrackTitle,
  bands: TrackBands,
})
export type TrackOfTrackCover = z.infer<typeof TrackOfTrackCover>
