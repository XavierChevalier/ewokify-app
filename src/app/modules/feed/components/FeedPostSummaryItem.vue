<script setup lang="ts">
  import AppUserAvatar from '@/app/components/AppUserAvatar.vue'
  import { localeFormatDistance } from '@/app/tools/date/Date'
  import { computed, PropType } from 'vue'
  import { Author } from '@/app/modules/feed/models/Author'
  import { isPropertyValid } from '@/app/tools/component-properties/PropertyValidator'
  import { AvailableFeedPostType } from '@/app/modules/feed/models/FeedPostType'

  const publicationDate = localeFormatDistance(new Date(), new Date(), {
    addSuffix: false,
  })

  const props = defineProps({
    author: {
      type: Object as PropType<Author>,
      required: true,
      validator: isPropertyValid(Author),
    },
    postType: {
      type: String as PropType<keyof typeof AvailableFeedPostType>,
      required: true,
      validator: (value: string) =>
        Object.keys(AvailableFeedPostType).includes(value),
    },
  })
  const content = computed(() => `${props.author.name} à partagé un morceau`)
</script>

<template>
  <div class="flex gap-3 items-center">
    <AppUserAvatar size="sm" :user-name="author.name" :url="author.avatar" />
    <span>{{ content }}</span>
    <div class="flex relative flex-shrink-0 gap-2 items-center">
      <div
        class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
      />
      <span class="text-neutral">{{ publicationDate }}</span>
    </div>
  </div>
</template>
