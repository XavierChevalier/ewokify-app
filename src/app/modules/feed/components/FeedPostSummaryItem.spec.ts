import { ComponentTestsGenerator } from '@/tests/tests-generators/ComponentTestsGenerator'
import FeedPostSummaryItem from './FeedPostSummaryItem.vue'
import { generateStorybookSnapshotTests } from '@/tests/tools/StorybookSnapshot'
import { Default } from '@/app/modules/feed/components/FeedPostSummaryItem.stories'
import { Author } from '@/app/modules/feed/models/Author'
import { omit } from 'lodash-es'
import { AvailableFeedPostType } from '@/app/modules/feed/models/FeedPostType'

describe('FeedPostSummaryItem', () => {
  generateStorybookSnapshotTests({ Default })

  const componentTestsGenerator = new ComponentTestsGenerator(
    FeedPostSummaryItem
  )
  componentTestsGenerator.itShouldBeDefined()

  const author = { name: 'Test', avatar: 'https://picsum.photos/50' }
  componentTestsGenerator
    .property<Author>('author')
    .itShouldBeDefined()
    .itShouldBeRequired()
    .itShouldHaveAValidatorFunction()
    .itShouldFailWith('missing name', omit(author, ['name']))
    .itShouldFailWith('a null avatar', { name: 'Test', avatar: null })
    .itShouldFailWith('a null name and avatar', { name: null, avatar: null })

  componentTestsGenerator
    .property<keyof typeof AvailableFeedPostType>('postType')
    .itShouldBeDefined()
    .itShouldBeTypeOf(String)
    .itShouldBeRequired()
    .itShouldHaveAValidatorFunction()
    .itShouldSuccessWith('newSingle')
    .itShouldSuccessWith('newAlbum')

  componentTestsGenerator.generateTests()
})
