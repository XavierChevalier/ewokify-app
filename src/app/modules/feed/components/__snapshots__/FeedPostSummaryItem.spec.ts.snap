// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedPostSummaryItem [snapshot] should render the "Default" component correctly 1`] = `
<div
  class="flex gap-3 items-center"
  publicationdate="Mon Nov 29 2021 01:00:00 GMT+0100 (heure normale d’Europe centrale)"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless à partagé un morceau
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;
